### PR TITLE
Ensure NSProcess exceptions are translated to Swift errors

### PR DESCRIPTION
**Pull request by werner77**
Use NSProcess.run instead of NSProcess.launch on supported mac version (10.13 and higher). This ensures that the calling application won't crash if for example the launchPath does not point to a valid executable.

### CODE REVIEW

File: Sources/Task.swift [hunk 1]
```diff
@@ -508,10 +508,22 @@ extension Task {
...
+								close(stdoutPipe.writeFD)
+								close(stderrPipe.writeFD)
+							}
+
+							if #available(macOS 10.13, *) {
+								try process.run()
+							} else {
+								process.launch()
+							}
+						} catch {
```
Comment (mdiep) [inline Sources/Task.swift]: Can you move this `do`/`catch` inside the `if` so that it wraps just the `try` statement?

IMO that makes the code clearer.

File: Sources/Errors.swift [hunk 2]
```diff
@@ -10,6 +10,10 @@ import Foundation
 
 /// An error originating from ReactiveTask.
 public enum TaskError: Error, Equatable {
+
+	/// A shell task failed to launch
+	case shellTaskLaunchFailed(Task, reason: String?)
```
Comment (mdiep) [inline Sources/Errors.swift]: This could probably be `launchFailed`. `shellTaskFailed` is meant to denote that the error comes from the shell task itself. In this case, the failure doesn't come from the shell task—it comes from the OS.

File: Tests/ReactiveTaskTests/TaskSpec.swift [hunk 3]
```diff
@@ -87,9 +87,34 @@ class TaskSpec: QuickSpec {
...
 			if let error = result.error {
-				expect(error.description) == "A shell task (/usr/bin/stat not-a-real-file) failed with exit code 1:\nstat: not-a-real-file: stat: No such file or directory\n"
+				let expectedDescription =
+					"A shell task (/usr/bin/stat not-a-real-file) failed with exit code 1:\n" +
+					"stat: not-a-real-file: stat: No such file or directory\n"
+				expect(error.description) == expectedDescription
+			}
+		}
+
+		it("should fail correctly with an invalid task") {
```
Comment (mdiep) [inline Tests/ReactiveTaskTests/TaskSpec.swift:103]: Thanks for adding a test! ✨

**Review by mdiep — CHANGES_REQUESTED** [review]

File: Sources/Task.swift (hunk 1 above)
Comment (werner77) [reply Sources/Task.swift]: changed it

File: Package.swift [hunk 4]
```diff
@@ -14,7 +14,7 @@ let package = Package(
     ],
     targets: [
         .target(name: "ReactiveTask", dependencies: ["Result", "ReactiveSwift"], path: "Sources"),
-        .testTarget(name: "ReactiveTaskTests", dependencies: ["ReactiveTask", "Quick", "Nimble"]),
+        .testTarget(name: "ReactiveTaskTests", dependencies: ["ReactiveTask", "Quick", "Nimble"])
```
Comment (mdiep) [inline Package.swift]: Please revert these changes. These trailing commas should not have been removed.

File: Sources/Task.swift [hunk 5]
```diff
@@ -518,11 +518,25 @@ extension Task {
...
-						process.launch()
-						close(stdoutPipe.writeFD)
-						close(stderrPipe.writeFD)
+
+						if #available(macOS 10.13, *) {
+							do {
+								defer {
+									close(stdoutPipe.writeFD)
+									close(stderrPipe.writeFD)
+								}
```
Comment (mdiep) [inline Sources/Task.swift:546]: This `defer` could probably move outside the `if` to cover both cases?

Comment (ikesyo) [reply Sources/Task.swift:546]: Is it okay the `defer` runs after the following block? (it may be no problem)

```swift
lifetime += stdinProducer.start()

lifetime.observeEnded {
	process.terminate()
}
```

Comment (mdiep) [reply Sources/Task.swift:546]: I don't think that'd be a problem?
